File: saksflyt/src/test/java/no/nav/melosys/saksflyt/steg/brev/BestillBrevTest.java
package no.nav.melosys.saksflyt.steg.brev;

import no.nav.melosys.domain.BehandlingTestFactory;
import no.nav.melosys.domain.brev.DoksysBrevbestilling;
import no.nav.melosys.domain.brev.Mottaker;
import no.nav.melosys.domain.kodeverk.Mottakerroller;
import no.nav.melosys.exception.FunksjonellException;
import no.nav.melosys.saksflyt.brev.BrevBestiller;
import no.nav.melosys.saksflytapi.domain.ProsessDataKey;
import no.nav.melosys.saksflytapi.domain.ProsessStatus;
import no.nav.melosys.saksflytapi.domain.ProsessType;
import no.nav.melosys.saksflytapi.domain.Prosessinstans;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.ArgumentCaptor;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;

import static no.nav.melosys.domain.kodeverk.brev.Produserbaredokumenter.INNVILGELSE_YRKESAKTIV;
import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThatExceptionOfType;
import static org.mockito.Mockito.verify;

@ExtendWith(MockitoExtension.class)
class BestillBrevTest {

    @Mock
    BrevBestiller brevBestiller;

    private BestillBrev bestillBrev;

    @BeforeEach
    void setUp() {
        bestillBrev = new BestillBrev(brevBestiller);
    }

    @Test
    void utfør_altOk_kallerBestill() {
        var behandling = BehandlingTestFactory.builderWithDefaults().build();
        DoksysBrevbestilling brevbestilling = new DoksysBrevbestilling.Builder()
            .medProduserbartDokument(INNVILGELSE_YRKESAKTIV)
            .medMottakere(Mottaker.medRolle(Mottakerroller.BRUKER))
            .build();
        var prosessinstans = Prosessinstans.builder()
            .medType(ProsessType.OPPRETT_SAK)
            .medStatus(ProsessStatus.KLAR)
            .medBehandling(behandling)
            .medData(ProsessDataKey.BREVBESTILLING, brevbestilling)
            .build();
        ArgumentCaptor<DoksysBrevbestilling> captor = ArgumentCaptor.forClass(DoksysBrevbestilling.class);


        bestillBrev.utfør(prosessinstans);


        verify(brevBestiller).bestill(captor.capture());
        DoksysBrevbestilling bestiltBrevbestilling = captor.getValue();
        assertThat(brevbestilling.getBehandling()).isNull();
        assertThat(bestiltBrevbestilling.getBehandling()).isEqualTo(behandling);
        assertThat(bestiltBrevbestilling.getProduserbartdokument()).isEqualTo(brevbestilling.getProduserbartdokument());
        assertThat(bestiltBrevbestilling.getMottakere()).isEqualTo(brevbestilling.getMottakere());
    }

    @Test
    void utfør_manglerBehandling_kasterFeilmelding() {
        var prosessinstans = Prosessinstans.builder().medType(ProsessType.OPPRETT_SAK).medStatus(ProsessStatus.KLAR).build();

        assertThatExceptionOfType(FunksjonellException.class)
            .isThrownBy(() -> bestillBrev.utfør(prosessinstans))
            .withMessageContaining("Prosessinstans mangler behandling");
    }

    @Test
    void utfør_manglerBrevbestilling_kasterFeilmelding() {
        var prosessinstans = Prosessinstans.builder()
            .medType(ProsessType.OPPRETT_SAK)
            .medStatus(ProsessStatus.KLAR)
            .medBehandling(BehandlingTestFactory.builderWithDefaults().build())
            .build();

        assertThatExceptionOfType(FunksjonellException.class)
            .isThrownBy(() -> bestillBrev.utfør(prosessinstans))
            .withMessageContaining("Prosessinstans mangler brevbestilling");
    }

    @Test
    void utfør_flereEnnEnMottaker_kasterFeilmelding() {
        var behandling = BehandlingTestFactory.builderWithDefaults().build();
        var prosessinstans = Prosessinstans.builder()
            .medType(ProsessType.OPPRETT_SAK)
            .medStatus(ProsessStatus.KLAR)
            .medBehandling(behandling)
            .medData(ProsessDataKey.BREVBESTILLING,
                new DoksysBrevbestilling.Builder()
                    .medMottakere(Mottaker.medRolle(Mottakerroller.BRUKER), Mottaker.medRolle(Mottakerroller.ARBEIDSGIVER))
                    .build())
            .build();

        assertThatExceptionOfType(FunksjonellException.class)
            .isThrownBy(() -> bestillBrev.utfør(prosessinstans))
            .withMessageContaining("Prosessinstans skal sende brev til én mottaker, fant 2");
    }
}

File: saksflyt/src/test/java/no/nav/melosys/saksflyt/steg/brev/BestillBrevTest.kt
package no.nav.melosys.saksflyt.steg.brev

import io.kotest.assertions.throwables.shouldThrow
import io.kotest.matchers.shouldBe
import io.kotest.matchers.string.shouldContain
import io.mockk.every
import io.mockk.junit5.MockKExtension
import io.mockk.mockk
import io.mockk.slot
import io.mockk.verify
import no.nav.melosys.domain.Behandling
import no.nav.melosys.domain.brev.DoksysBrevbestilling
import no.nav.melosys.domain.brev.Mottaker
import no.nav.melosys.domain.forTest
import no.nav.melosys.domain.kodeverk.Mottakerroller
import no.nav.melosys.domain.kodeverk.brev.Produserbaredokumenter.INNVILGELSE_YRKESAKTIV
import no.nav.melosys.exception.FunksjonellException
import no.nav.melosys.saksflyt.brev.BrevBestiller
import no.nav.melosys.saksflytapi.domain.ProsessDataKey
import no.nav.melosys.saksflytapi.domain.Prosessinstans
import no.nav.melosys.saksflytapi.domain.forTest
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.extension.ExtendWith

@ExtendWith(MockKExtension::class)
class BestillBrevTest {

    private val brevBestiller: BrevBestiller = mockk()

    private lateinit var bestillBrev: BestillBrev

    @BeforeEach
    fun setUp() {
        bestillBrev = BestillBrev(brevBestiller)
    }

    @Test
    fun utfør_altOk_kallerBestill() {
        val behandling = Behandling.forTest { }
        val prosessinstans = Prosessinstans.forTest {
            this.behandling = behandling
        }
        val brevbestilling = DoksysBrevbestilling.Builder()
            .medProduserbartDokument(INNVILGELSE_YRKESAKTIV)
            .medMottakere(Mottaker.medRolle(Mottakerroller.BRUKER))
            .build()
        prosessinstans.setData(ProsessDataKey.BREVBESTILLING, brevbestilling)
        val slot = slot<DoksysBrevbestilling>()
        every { brevBestiller.bestill(capture(slot)) } returns Unit


        bestillBrev.utfør(prosessinstans)


        verify { brevBestiller.bestill(any()) }
        val bestiltBrevbestilling = slot.captured
        brevbestilling.behandling shouldBe null
        bestiltBrevbestilling.run {
            this.behandling shouldBe behandling
            produserbartdokument shouldBe brevbestilling.produserbartdokument
            mottakere shouldBe brevbestilling.mottakere
        }
    }

    @Test
    fun utfør_manglerBehandling_kasterFeilmelding() {
        val prosessinstans = Prosessinstans()


        val exception = shouldThrow<FunksjonellException> {
            bestillBrev.utfør(prosessinstans)
        }


        exception.message shouldContain "Prosessinstans mangler behandling"
    }

    @Test
    fun utfør_manglerBrevbestilling_kasterFeilmelding() {
        val prosessinstans = Prosessinstans.forTest {
            behandling = Behandling.forTest { }
        }


        val exception = shouldThrow<FunksjonellException> {
            bestillBrev.utfør(prosessinstans)
        }


        exception.message shouldContain "Prosessinstans mangler brevbestilling"
    }

    @Test
    fun utfør_flereEnnEnMottaker_kasterFeilmelding() {
        val behandling = Behandling.forTest { }
        val prosessinstans = Prosessinstans.forTest {
            this.behandling = behandling
            medData(
                ProsessDataKey.BREVBESTILLING,
                DoksysBrevbestilling.Builder()
                    .medMottakere(Mottaker.medRolle(Mottakerroller.BRUKER), Mottaker.medRolle(Mottakerroller.ARBEIDSGIVER))
                    .build()
            )
        }


        val exception = shouldThrow<FunksjonellException> {
            bestillBrev.utfør(prosessinstans)
        }


        exception.message shouldContain "Prosessinstans skal sende brev til én mottaker, fant 2"
    }
}
